STREAMS-1569 - Database guidelines (#57)
Axway/streams-open-docs: content/en/docs/architecture/_index.md

## Overview

Streams is an event hub that makes it easy to exchange messages between devices, services and applications. It only supports container-based deployment. The purpose of this document is to share Axway reference architecture for the container-based deployment of a Streams solution on Kubernetes. It will address many architectural, development and operational aspects of the proposed architecture.  

> Because the technology chosen, Docker and Kubernetes, are portable across on-premise environments and many cloud providers, most of the information in this guide should apply to those environments. But we include specific recommendations for AWS as one of the most common deployment targets.

Reviewer: ```suggestion
Because Docker and Kubernetes are portable across on-premise environments and many cloud providers, most of the information in this guide should apply to those environments. However, we've also included specific recommendations for AWS, as it is one of the most common deployment targets.
```

### Load balancer

> A load balancer is required in front of the cluster. We use the Kubernetes object called ingress controller that is responsible for fulfilling the ingress rules. A NGINX ingress controller is, by default, deployed with our Helm chart installation. For better performance, we recommend you to use a L4 Load balancer. See your Cloud provider documentation to configure it (e.g. [AWS Load Balancing](https://docs.aws.amazon.com/eks/latest/userguide/load-balancing.html))

Reviewer: ```suggestion
A load balancer is required in front of the cluster. We use the Kubernetes object called ingress controller that is responsible for fulfilling the ingress rules. A NGINX ingress controller is, by default, deployed with our Helm chart installation. For better performance, we recommend you to use a L4 Load balancer. See your Cloud provider documentation to configure it (e.g. [AWS Load Balancing](https://docs.aws.amazon.com/eks/latest/userguide/load-balancing.html)).
```

> Kubernetes Role-based access control ([RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)) is a method of regulating access to your Kubernetes cluster and resources based on the roles of individual users within an enterprise. Streams requires RBAC to be enabled for secrets management and third-party dependencies:

Reviewer: ```suggestion
Kubernetes' role-based access control ([RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)) is a method of regulating access to your Kubernetes cluster and resources based on the roles of individual users within an enterprise. Streams requires RBAC to be enabled for secrets management and third-party dependencies:
```

  application: streams
```

> For 3rd parties pods (NGINX, Kafka, Zookeeper and MariaDB) you have to define it for each of them:

Reviewer: ```suggestion
You must define it for each of the 3rd party pods (NGINX, Kafka, Zookeeper and MariaDB):
```

##### External traffic

> We use the NGINX Ingress Controller to expose both management and subscription APIs. NGINX Ingress Controller automatically creates a Load Balancer in the cloud provider infrastructure so that the APIs can be reached from the outside using the load balancer DNS hostname. Once known, this hostname must be updated in the ingress resource thanks to the value of the helm parameter `ingress.host`.

Reviewer: ```suggestion
We use the NGINX Ingress Controller to expose both management and subscription APIs. NGINX Ingress Controller automatically creates a load balancer in the cloud provider infrastructure so that the APIs can be reached from the outside using the load balancer DNS hostname. Once known, this hostname must be updated in the ingress resource thanks to the value of the helm parameter `ingress.host`.
```

Reviewer: Is the hostname updated in parameter `ingress.host`? If so, I recommend: Once known, this hostname must be updated in the ingress resource helm parameter `ingress.host`.

> SSL/TLS is enabled by default on NGINX ingress controller unless you have explicitly disabled it (`ingress.tlsenabled=false`).

Reviewer: ```suggestion
SSL/TLS is enabled by default on NGINX Ingress Controller unless you have explicitly disabled it (`ingress.tlsenabled=false`).
```

> The ingress controller handles SSL/TLS termination with to the correct certificate/secret (more setup details)

Reviewer: ```suggestion
The ingress controller handles SSL/TLS termination with the correct certificate/secret (more setup details).
```

Reviewer: I don't understand the following: (more setup details). Is this an internal note meaning that more setup details should be provided by the author?

Reviewer 2: @jfmarquet @benjaminranc Need your input on this as I do not understand it neither.

##### External traffic

We use the NGINX Ingress Controller to expose both management and subscription APIs. NGINX Ingress Controller automatically creates a load balancer in the cloud provider infrastructure so that the APIs can be reached from the outside using the load balancer DNS hostname. Once known, this hostname must be updated in the ingress resource helm parameter `ingress.host`.

SSL/TLS is enabled by default on NGINX Ingress Controller unless you have explicitly disabled it (`ingress.tlsenabled=false`).

> The ingress controller handles SSL/TLS termination with a certificate manage with a k8s secret (refer to the [Ingress TLS settings](/docs/install#ingress-tls-settings) documentation for further details).

Reviewer 2: ```suggestion
The ingress controller handles the SSL/TLS termination thanks to a certificate stored in a k8s secret (refer to the [Ingress TLS settings](/docs/install#ingress-tls-settings) documentation for further details).
```

Here is the list of secrets related to Streams installation:

| Description                   | Name                      | Type                                |
| ----------------------------- | ------------------------- | ----------------------------------- |
| docker registry credentials   | registry-secret-name      | kubernetes.io/dockerconfigjson      |
| mariadb credentials           | database-passwords-secret | Opaque                              |
| mariadb encryption            | database-secret           | Opaque                              |
| kafka truststore credentials  | client-jks-secret         | Opaque                              |
| kafka credentials             | passwords-secret          | Opaque                              |
| kafka service account         | kafka-token               | kubernetes.io/service-account-token |
| Nginx tls certificates        | ingress-tls-secret        | kubernetes.io/tls                   |
| Nginx service account         | ingress-nginx-token       | kubernetes.io/service-account-token |

> | helm release internal info    | helm                      | helm.sh/release.v1                  |

Reviewer 3: It seeems that most of the secret name are invalid and are missing the prefix 'streams-'

- streams-database-passwords-secret
- streams-database-secret
- streams-kafka-passwords-secret
- streams-kafka-client-jks-secret
- streams-ingress-tls-secret

Without secrets, all passwords are set in clear in Manifest. Kubernetes define “secret” objects to encode in base64 all sensitive information. Using Kubernetes Secrets is very useful for variables in containers, Docker registry login, and technical token for shared storage.

> Here is the list of secrets related to Streams installation:

Reviewer: ```suggestion
The following list of secrets relate to the Streams installation:
```

{{< alert title="Note" >}}MariaDB is deployed in master/slave mode with asynchronous commit for replication but the failover is not done automatically.{{< /alert >}}

#### NGINX

NGINX is the ingress controller in front of Streams APIs.

Source Docker image:

* Repository: k8s.gcr.io/ingress-nginx/controller
* Tag: v0.43.0

Pod name: `streams-ingress-nginx-controller`

Pod characteristics for HA deployment mode:

| Replicas           | CPU | Memory | Xms & Xmx | Persistence |
| ------------------ | --- | ------ | --------- | ----------- |
| 2                  | 1   | 512 MB | n/a       | 8 GB        |

### Database considerations

Streams manages 2 types of connection pool for the database:

* Tomcat:
    * At several microservices startup, schema database management ([liquibase](https://www.liquibase.org/)) creates a connection pool with 10 parallel connections. Using these connections (with user _root_), database tables are created/updated if needed in order to reach the expected state for the current version. Then, the connections are closed and the microservices can resume regular startup.
* [Hikari](https://github.com/brettwooldridge/HikariCP):

>     * Each microservice which depends on the database maintains a connection pool of `maximumPoolSize` threads (default: 10). Connections are established using user _streams_. When a database call needs to be performed (e.g. topic creation, liveness probe...), either an existing idle connection is available and used for this call or, all the connections in the pool are already in use and the call is queued during `connectionTimeout` (default: 25s).

Reviewer: ```suggestion
    * Each microservice, which depends on the database, maintains a connection pool of `maximumPoolSize` threads (default: 10). Connections are established using user _streams_. When a database call needs to be performed (e.g. topic creation, liveness probe...), either an existing idle connection is available and used for this call, or all the connections in the pool are already in use and the call is queued during `connectionTimeout` (default: 25s).
```

    * A connection remains in the pool during `maxLifetime` (default: 280s). When maxLifetime is reached and if the connection is not in use, it will be dropped from the pool and a new one will be created in order to keep the pool size at `maximumPoolSize` connections.

> For best performances with Streams, MariaDB should be configured as follows:

Reviewer: ```suggestion
For the best performance with Streams, MariaDB should be configured as follows:
```

* [`wait-timeout`](https://mariadb.com/docs/reference/es/system-variables/wait_timeout/)
    * Any connection to the database that stays idle during this value will be destroyed.

>     * 5 minutes is a good value, so reconnection are not too frequent and zombie connections due to potential k8s pod or node crash can be managed fast enough.

Reviewer: ```suggestion
    * 5 minutes is recommended so that reconnections are not too frequent and zombie connections that are due to potential k8s pod / node crashes can be managed in a timely manner.
```

>     * Keep in mind to set your Streams installation `maxLifetime` below this value. For instance, a microservice connection will die after remaining `maxLifetime` seconds in the pool, but if a pod crashes and the connection cannot be closed properly after `maxLifetime`, it will eventually be destroyed after `wait-timeout`.

Reviewer: ```suggestion
    * Remember to set your Streams installation `maxLifetime` below this value. For instance, a microservice connection will die after remaining `maxLifetime` seconds in the pool, but if a pod crashes and the connection cannot be closed properly after `maxLifetime`, it will eventually be destroyed after `wait-timeout`.
```

* [`max-connections`](https://mariadb.com/docs/reference/es/system-variables/max_connections/)
    * This value is set to 500 by default for HA setup.
    * The formula to compute the number of connections maintained by Streams platform is:
    ```
    streams_db_connections = (number_of_pods_depending_on_db) * (maximumPoolSize)
    
    Example:
    Assuming a platform deployed with hub, webhook subscriber, http post publisher with 2 replicas each:
    number_of_pods_depending_on_db = (hub*2+post*2+webhook*2) = 6
    streams_db_connections = 6 * 10 = 60
    ```

>     * One may also take into account the load expected on the platform which may lead to increase the number of microservice replicas. It is recommended to compute the `streams_db_connections` with the highest number of pod replicas expected for the platform to handle high load.

Reviewer 2: ```suggestion
    * One may also take into account the expected load on the platform, which may lead to an increase in the number of microservice replicas. It is recommended to compute the `streams_db_connections` with the highest number of pod replicas expected so that the platform can properly handle peak load.
```

>     * One may also take into account the expected load on the platform, which may lead to an increase in the number of microservice replicas. It is recommended to compute the `streams_db_connections` with the highest number of pod replicas expected so that the platform can properly handle peak load.

Reviewer: ```suggestion
    * Also take into account the expected load on the platform, which may lead to an increase in the number of microservice replicas. It is recommended to compute the `streams_db_connections` with the highest number of pod replicas expected so that the platform can properly handle peak load.
```

>     Several connections are also established by MariaDB itself (5 for InnoDB and 1 for replication).

Reviewer: ```suggestion
    Several connections are also established by MariaDB (five for InnoDB and one for replication).
```
